### PR TITLE
Extend minSizeInBits to cover more cases

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Fixes https://github.com/p4lang/p4-spec/issues/937

### CODE REVIEW

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:4783]: Should this be `e.minSizeInBits()`?

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:4790]: Do we want the max/min semantics here?

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:4786]: Do we want this semantics? And is it consistent with what we do for header unions and stacks below?

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:4796]: Do we want to enforce stricter information hiding for `newtype`?

**Comment by jnfoster** [comment]
We discussed this issue at the October '21 LDWG. 

There is broad support for this proposal, but some questions about certain types (header stacks, unions, etc.). 

We resolved to remove this operation for those types.

Action items:
- @vgurevich should weigh in on the proposal
- @mbudiu-vmw will modify the proposal to implement the above changes

File: p4-16/spec/P4-16-spec.mdk
Comment (vgurevich) [reply p4-16/spec/P4-16-spec.mdk:4786]: I can see the consistency with the definition of `minSizeInBits()` for the headers above (where we state that varbits are counted as 0).

However, for the practical purposes, it would be a lot more useful to have at least **some way** to return `N` for a `varbit<N>`.

One way to do that, might be to state that `e.minSizeInBits()` is really the minimum number of bits that should be allocated to store ANY value of the corresponding type. This, by the way, is consistent with what we do with header stacks that can also be only partially filled. This also works well with the idea that we allocate all resources statically and as a result there are no exceptions in the controls.

File: p4-16/spec/P4-16-spec.mdk
Comment (vgurevich) [reply p4-16/spec/P4-16-spec.mdk:4796]: @jnfoster -- I do not think we need to hide the size of the `newtype`, becaue it still ahs to be visible via table keys, action parameters, etc. The only thing we really want to do (in that specific case where the type is used for port (re)mapping is to constrain some operations that might be dangerous (e.g. ternary matching, math, etc.). Note by the way that today the constraints are so severe that this construct is almost useless.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:4178]: Proposal is for this to be 0.

**Review by jnfoster — APPROVED** [review]
Looks good. I like the way you converted the Google Doc notation.

Minor: maybe we should split out controls/parsers/packages?

**Comment by mbudiu-vmw** [comment]
Was actually thinking to take out all the undefined ones and say that everything is is undefined...

**Comment by jnfoster** [comment]
That could work too. Probably more robust that way.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:8135]: looks like some kind of typo here.  Perhaps remove "the of"?

**Comment by jnfoster** [comment]
Any the double “to apply”

**Comment by jfingerh** [comment]
Would it be reasonable to include at least a link to another file inside this same repo that gives the rationale that Nate recorded in his notes, for the cases like header stacks and header unions that one might consider "odd"?  Also the example Nate wrote that seems counter-intuitive when written out and demonstrated how minSizeInBits and maxSizeInBits behave on those data types?  I understand if we don't want to include such background material in the spec itself, but having a separate MarkDown file in the directory `p4-16/spec/docs` has been done for similar background details for other things in the spec, and seems useful as history of "why".

**Comment by jnfoster** [comment]
I personally think that would be fine...

-N

**Comment by jafingerhut** [comment]
I am happy adding the separate document with background details/rationale in a separate PR after this one is merged, if that helps this one make progress.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:4762]: typo: duplicated "the result the result"

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:4767]: Same typo "the result the result" here.

**Review by jafingerhut — APPROVED** [review]
Looks good to me, with only minor comments about a couple of typos.  I am fine if we wait for a later PR to add a separate file with background details/rationale.

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:8135]: Capitalize the "e" here.

**Comment by jnfoster** [comment]
We agreed to take this as written (modulo typos) at December '21 meeting.

**Pull request merged**
